lib/vrl/compiler/src/expression/assignment.rs

```diff
@@ -178,6 +208,100 @@ fn verify_mutable(
     }
 }
 
+/// Ensure that the given target is allowed to be changed.
+///
+/// This returns an error if an assignment is done to an object field or array
+/// index, while the parent of the field/index isn't an actual object/array.
+fn verify_overwriteable(
+    target: &Target,
+    local: &LocalEnv,
+    external: &ExternalEnv,
+    target_span: Span,
+    expr_span: Span,
+    assignment_span: Span,
+    rhs_expr: Expr,
+) -> Result<(), Error> {
+    let mut path = target.lookup();
+
+    let root_kind = match target {
+        Target::Noop => Kind::any(),
+        Target::Internal(ident, _) => local
+            .variable(ident)
+            .map(|detail| detail.type_def.kind().clone())
+            .unwrap_or_else(Kind::any),
+        Target::External(_) => external.target_kind().clone(),
+    };
+
+    let mut parent_span = target_span;
+    let mut remainder_str = String::new();
+
+    // Walk the entire path from back to front. If the popped segment is a field
+    // or index, check the segment before it, and ensure that its kind is an
+    // object or array.
+    while let Some(last) = path.pop_back() {
+        let parent_kind = root_kind
+            .find_at_path(&path)
+            .ok()
+            .flatten()
+            .map(|kind| kind.into_owned())
+            .unwrap_or_else(Kind::any);
+
+        let (variant, segment_span, valid) = match last {
+            segment @ Segment::Field(_) | segment @ Segment::Coalesce(_) => {
+                let segment_str = segment.to_string();
+                let segment_start = parent_span.end() - segment_str.len();
+                let segment_span = Span::new(segment_start, parent_span.end());
+
+                parent_span = Span::new(parent_span.start(), segment_start - 1);
+                remainder_str.insert_str(0, &format!(".{}", segment_str));
+
+                ("object", segment_span, parent_kind.contains_object())
+            }
+            Segment::Index(index) => {
+                let segment_start = parent_span.end() - format!("[{index}]").len();
+                let segment_span = Span::new(segment_start, parent_span.end());
+
+                parent_span = Span::new(parent_span.start(), segment_start);
+                remainder_str.insert_str(0, &format!("[{index}]"));
+
+                ("array", segment_span, parent_kind.contains_array())
+            }
+        };
+
+        if valid {
+            continue;
+        }
+
+        let parent_str = match target {
+            Target::Internal(ident, _) => format!("{ident}{}", path.to_string()),
+            Target::External(_) => {
+                if path.is_root() && remainder_str.starts_with('.') {
+                    remainder_str = remainder_str[1..].to_owned();
+                }
+
+                format!(".{}", path.to_string())
+            }
+            Target::Noop => unreachable!(),
+        };
```

Comment:
Displaying paths is — and always has been — gnarly. I considered fixing this at the root (e.g. `Lookup(Buf)`), but decided against it, since it would require changing numerous other locations where we do hacks similar to this, and I'd rather do that change in one go, when we move to "lookup v2" proper.